ocaml/typing/typedecl.ml

```diff
@@ -2306,6 +2315,27 @@ let check_unboxable env loc ty =
     all_unboxable_types
     ()
 
+let prim_expecting_any prim =
+  match prim.prim_name with
+  | "%array_length"
+  | "%array_safe_get"
+  | "%array_safe_set"
+  | "%array_unsafe_get"
+  | "%array_unsafe_set" -> false
+  | _ -> true
+
+let error_if_jkind_any_occurs_unexpectly prim env cty ty =
+  if prim.prim_is_layout_representation_polymorphic then ()
```

Comment:
There are a lot of primitives, only some of which actually support `rep_poly`, I think.  For example, this crashes the compiler late (in flambda2):
```
external[@rep_poly] dup : ('a : any). 'a -> 'a = "%obj_dup"

let x =
  let f = dup #3.14 in
  if Stdlib__Float_u.(compare f #2.0) > 0 then 1 else 2
```
And I bet we can come up with similar examples that generate code that just segfaults.  Scary!

As I noted elsewhere, there is not a lot safety checking around primitives, so this isn't so far out of the ordinary.  But I think it will be pretty hard for users to predict which primitives do and which don't support `rep_poly`, and that set will change over time.  Would it be easy to provide some static checking by just making a list here of the ones we know to work and erroring if the user has asked for `rep_poly` but the primitive doesn't support it?

Reply:
Think the issue pointed out here is not specific to `rep_poly` but rather about non-value jkinds more generally.

```
external dup : float# -> float# = "%obj_dup"
```

has the same problem as the example given.

I reworked the checks and documented them here:

https://github.com/ocaml-flambda/flambda-backend/blob/9e25a6573a24a00c911b2a1b4add9990205549a9/ocaml/typing/typedecl.ml#L2367-L2400

There's now a list of primitives that are allowed to have argument/return types with non-value jkinds and the example given will result in an error after the change.